relay-quotas/src/redis.rs

```diff
@@ -125,6 +139,291 @@ impl std::ops::Deref for RedisQuota<'_> {
+}
+
+impl Default for GlobalCounters {
+    fn default() -> Self {
+        Self {
+            script: Arc::new(load_global_lua_script()),
+            counters: Default::default(),
+        }
+    }
+}
+
+/*
+impl fmt::Debug for GlobalCounters {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let mut debug_struct = f.debug_struct("GlobalCounters");
+
+        if let Ok(counters) = self.counters.read() {
+            for (key, val) in counters.iter() {
+                let budget = val.budget.load(Ordering::SeqCst);
+                let slot = val.slot;
+                let redis = val.last_seen_redis_value;
+
+                debug_struct.field(
+                    &format!("{:?}", key),
+                    &format!("budget: {}, redis_value: {}, slot {}", budget, redis, slot),
+                );
+            }
+        } else {
+            debug_struct.field("counters", &"Lock Poisoned");
+        }
+
+        debug_struct.finish()
+    }
+}
+*/
+
+impl GlobalCounters {
+    /// Returns `true` if the global quota should be ratelimited.
+    ///
+    /// Certain errors can be resolved by syncing to redis, so in those cases
+    /// we try again to decrement the budget after syncing.
+    fn evaluate_quota(
+        &self,
+        client: &mut PooledClient,
+        quota: &RedisQuota,
+        quantity: usize,
+    ) -> Result<bool, GlobalRateLimitError> {
+        use GlobalRateLimitError as E;
+
+        match self.is_rate_limited(quota, quantity) {
+            ok @ Ok(_) => ok,
+            err @ Err(E::Redis) => err,
+            Err(E::BudgetEmpty | E::SlotExpired | E::KeyMissing) => {
+                self.redis_sync(client, quota, quantity)?;
+                self.is_rate_limited(quota, quantity)
+            }
+        }
+    }
+
+    /// Returns `true` if items should be ratelimited.
+    fn is_rate_limited(
+        &self,
+        quota: &RedisQuota,
+        quantity: usize,
+    ) -> Result<bool, GlobalRateLimitError> {
+        let Some(limit) = quota.limit.map(|limit| limit as usize) else {
+            return Ok(false);
+        };
+
+        let current_slot = current_slot(quota.window) as usize;
+        let map = self.counters.read().unwrap_or_else(|e| e.into_inner());
+        let val = map
+            .get(&BudgetKeyRef::new(quota))
+            .ok_or(GlobalRateLimitError::KeyMissing)?
+            .read()
+            .unwrap_or_else(|e| e.into_inner());
+
+        if current_slot != val.slot {
+            return Err(GlobalRateLimitError::SlotExpired);
+        }
+
+        let redis_count = val.last_seen_redis_value;
+
+        loop {
+            let current_budget = val.budget.load(Ordering::SeqCst);
+            let total_count = redis_count.saturating_sub(current_budget);
+
+            if total_count + quantity > limit {
+                return Ok(true);
+            }
+
+            if current_budget < quantity {
+                return Err(GlobalRateLimitError::BudgetEmpty);
```

Comment:
so yes, this would indeed make it more simple, although with a decent performance loss. Looking forward to the irl discussion on whether it's worth it or not. 
 
About removing the BudgetEmpty and the KeyMissing (as mentioned in other place), i dont see the point of that, they're essentially just there for readability as keymissing/budgetempty/slotexceeded are all fixed by syncing to redis. so it's not really 'state' in practice